File: backend/ai.py
"""
AI answer step for the Ministry of Climate Change, Environment and Energy
(Maldives) RTI portal.

Combines two grounding sources:

    1. RAG over the ministry's local archive — past responded RTI requests +
       FAQs. Retrieved items are injected into the system prompt as a
       "Ministry archive" block.
    2. Live web search and fetch via Claude's server-side tools, restricted
       to rtidhonbe.com (preferred) and environment.gov.mv (fallback).

The model gets ministry precedent (RAG) plus current facts (web) and is
instructed to ground every claim in one of them.
"""

from __future__ import annotations

import logging
import os
from typing import Optional

import anthropic

from rag import RAGIndex, format_retrieved_for_prompt

log = logging.getLogger(__name__)

_MODEL = "claude-haiku-4-5"
_MAX_TOKENS = 2048
_MAX_PAUSE_TURNS = 3

_SYSTEM_PROMPT_TEMPLATE = (
    "You are an AI assistant for the Maldives Ministry of Climate Change, "
    "Environment and Energy's citizen Right to Information (RTI) portal.\n\n"
    "When a citizen submits an RTI request, your job is to draft a clear, "
    "factual response addressed to the citizen, grounded in two sources:\n\n"
    "A. THE MINISTRY ARCHIVE — past responded RTI requests and standing FAQs, "
    "   retrieved for you and shown below. These are authoritative precedent "
    "   and process knowledge. PREFER them when they answer the question.\n\n"
    "B. LIVE OFFICIAL SOURCES — accessible via the web_search and web_fetch "
    "   tools, restricted to:\n"
    "     1. rtidhonbe.com (the RTI vault) — search this FIRST.\n"
    "     2. environment.gov.mv (the ministry site) — search this ONLY IF the "
    "        vault does not have what is needed.\n\n"
    "DECISION ORDER:\n"
    "1. If the archive items below directly answer the question, draft from "
    "   them and cite the prior RTI id (e.g. RTI-2024-0001) or FAQ id.\n"
    "2. Otherwise use the web tools, vault first then the ministry site.\n"
    "3. If neither has the answer, say so plainly and direct the citizen to "
    "   file a formal RTI application with the Information Officer.\n\n"
    "MINISTRY ARCHIVE:\n{archive_block}\n\n"
    "RULES:\n"
    "- Every factual claim must come from the archive or from content you "
    "  retrieved via web_search/web_fetch. Do not invent figures, names, "
    "  dates, or document references.\n"
    "- State which source(s) you used (archive precedent, FAQ, rtidhonbe.com, "
    "  or environment.gov.mv).\n"
    "- Address the citizen directly. Be concise: 4-8 sentences, plain prose, "
    "  no markdown headings.\n\n"
    "OUTPUT FORMAT:\n"
    "Your final text reply will be shown to the citizen verbatim. Do NOT include "
    "search narration (\"I'll search...\", \"Let me try...\", \"The search "
    "returned...\"). Do NOT include any preamble or signoff like \"Response to "
    "Citizen:\", \"Dear Citizen,\", or \"Best regards\". Output ONLY the body of "
    "the response, in plain prose, starting with the substantive answer."
)

_ALLOWED_DOMAINS = ["rtidhonbe.com", "environment.gov.mv"]

_TOOLS = [
    {
        "type": "web_search_20260209",
        "name": "web_search",
        "allowed_domains": _ALLOWED_DOMAINS,
        "max_uses": 6,
        # Haiku 4.5 does not support programmatic tool calling (the default
        # mode for the _20260209 web tools). "direct" invokes the tool the
        # classic way (Claude emits a tool_use block) and disables dynamic
        # result filtering.
        "allowed_callers": ["direct"],
    },
    {
        "type": "web_fetch_20260209",
        "name": "web_fetch",
        "allowed_domains": _ALLOWED_DOMAINS,
        "max_uses": 8,
        "allowed_callers": ["direct"],
    },
]


def answer_request(
    *,
    subject: str,
    description: str,
    rag_index: Optional[RAGIndex] = None,
    rag_k: int = 4,
) -> str:
    """
    Generate a citizen-facing response, grounded in the ministry archive (RAG)
    and live web sources. Raises RuntimeError on hard failures; returns a
    clearly-labelled stub if ANTHROPIC_API_KEY is unset.
    """
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        log.warning("ANTHROPIC_API_KEY not set; returning stub response.")
        return (
            "[AI service not configured: set ANTHROPIC_API_KEY] "
            "Your request to the Ministry of Climate Change, Environment and "
            "Energy has been received and is pending review."
        )

    # RAG retrieval — pulls past responded RTIs + relevant FAQs as precedent.
    archive_hits: list[dict] = []
    if rag_index is not None:
        query = f"{subject}\n{description}".strip()
        archive_hits = rag_index.retrieve(query, k=rag_k)
    system_prompt = _SYSTEM_PROMPT_TEMPLATE.format(
        archive_block=format_retrieved_for_prompt(archive_hits)
    )

    user_prompt = (
        f"Subject: {subject}\n"
        f"Description: {description}\n\n"
        "First check the ministry archive shown in your instructions. If the "
        "archive does not answer the question, look up rtidhonbe.com (then "
        "environment.gov.mv). Then draft the response to the citizen."
    )

    client = anthropic.Anthropic(api_key=api_key)
    messages: list[dict] = [{"role": "user", "content": user_prompt}]

    for _ in range(_MAX_PAUSE_TURNS + 1):
        message = client.messages.create(
            model=_MODEL,
            max_tokens=_MAX_TOKENS,
            system=system_prompt,
            tools=_TOOLS,
            messages=messages,
        )
        if message.stop_reason != "pause_turn":
            break
        messages.append({"role": "assistant", "content": message.content})
    else:
        raise RuntimeError("Web-tool loop did not converge within pause_turn cap.")

    # When server-side tools run, the model emits short text blocks between
    # tool calls ("I'll search the vault...", "Let me try again..."). Those
    # are planning narration, not the answer. Keep only text blocks that
    # appear AFTER the last tool-use block in the content array.
    last_tool_idx = -1
    for i, block in enumerate(message.content):
        if block.type in {"tool_use", "server_tool_use"}:
            last_tool_idx = i

    final_text_parts = [
        block.text.strip()
        for block in message.content[last_tool_idx + 1 :]
        if block.type == "text" and block.text.strip()
    ]
    if not final_text_parts:
        raise RuntimeError("Anthropic response contained no final text block.")
    return "\n\n".join(final_text_parts)

File: backend-go/internal/ai/ai.go
// Package ai talks to the Anthropic Messages API to (a) structure a freshly
// filed citizen request for officer review and (b) draft a citizen-facing
// response grounded in the ministry archive.
//
// The Go port uses the Anthropic REST API directly over net/http — no SDK —
// to keep the dependency surface tiny.
//
// Web search/fetch tools (used by the Python AI step) are intentionally NOT
// wired up here. They are an optional grounding source; the local archive
// retrieval (RAG + graph) is the primary signal and produces high-quality
// drafts on its own.
package ai

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/rti4all/backend-go/internal/graph"
	"github.com/rti4all/backend-go/internal/rag"
)

// Client is a thin Anthropic Messages-API client.
type Client struct {
	APIKey     string
	Model      string
	HTTPClient *http.Client
}

// New constructs an AI client. An empty API key is allowed — calls will
// short-circuit with a stub response, mirroring the Python behaviour.
func New() *Client {
	model := os.Getenv("ANTHROPIC_MODEL")
	if model == "" {
		// claude-haiku-4-5 was the Python default; pin to the same model.
		model = "claude-haiku-4-5"
	}
	return &Client{
		APIKey:     strings.TrimSpace(os.Getenv("ANTHROPIC_API_KEY")),
		Model:      model,
		HTTPClient: &http.Client{Timeout: 60 * time.Second},
	}
}

// Configured reports whether the client has a usable API key.
func (c *Client) Configured() bool {
	return c.APIKey != "" && !strings.HasPrefix(c.APIKey, "sk-ant-placeholder")
}

// ----------------------------------------------------------------------------
// Anthropic API plumbing
// ----------------------------------------------------------------------------

type apiMessageRequest struct {
	Model     string       `json:"model"`
	MaxTokens int          `json:"max_tokens"`
	System    string       `json:"system,omitempty"`
	Messages  []apiMessage `json:"messages"`
}

type apiMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type apiMessageResponse struct {
	ID      string `json:"id"`
	Type    string `json:"type"`
	Role    string `json:"role"`
	Content []struct {
		Type string `json:"type"`
		Text string `json:"text"`
	} `json:"content"`
	StopReason string `json:"stop_reason"`
	Error      *struct {
		Type    string `json:"type"`
		Message string `json:"message"`
	} `json:"error,omitempty"`
}

func (c *Client) callMessages(ctx context.Context, req apiMessageRequest) (string, error) {
	if !c.Configured() {
		return "", fmt.Errorf("anthropic API key not configured")
	}
	body, err := json.Marshal(req)
	if err != nil {
		return "", fmt.Errorf("marshal request: %w", err)
	}
	httpReq, err := http.NewRequestWithContext(ctx, "POST", "https://api.anthropic.com/v1/messages", bytes.NewReader(body))
	if err != nil {
		return "", fmt.Errorf("build request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	httpReq.Header.Set("x-api-key", c.APIKey)
	httpReq.Header.Set("anthropic-version", "2023-06-01")

	resp, err := c.HTTPClient.Do(httpReq)
	if err != nil {
		return "", fmt.Errorf("anthropic call: %w", err)
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("read response: %w", err)
	}
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("anthropic returned HTTP %d: %s", resp.StatusCode, string(raw))
	}
	var parsed apiMessageResponse
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return "", fmt.Errorf("decode response: %w", err)
	}
	if parsed.Error != nil {
		return "", fmt.Errorf("anthropic error: %s: %s", parsed.Error.Type, parsed.Error.Message)
	}
	var text strings.Builder
	for _, block := range parsed.Content {
		if block.Type == "text" {
			if text.Len() > 0 {
				text.WriteString("\n\n")
			}
			text.WriteString(strings.TrimSpace(block.Text))
		}
	}
	return text.String(), nil
}

// ----------------------------------------------------------------------------
// Drafting a citizen-facing response
// ----------------------------------------------------------------------------

const answerSystemTemplate = `You are an AI assistant for the Maldives Ministry of Climate Change, Environment and Energy's citizen Right to Information (RTI) portal.

When a citizen submits an RTI request, your job is to draft a clear, factual response addressed to the citizen, grounded in the ministry's local archive of past responded RTI requests and standing FAQs.

The archive is retrieved for you two ways and shown below:
  - Vector matches: items semantically similar to the question.
  - Graph-linked items: items that share key concepts with the question.

These are authoritative precedent and process knowledge. PREFER them when they answer the question.

If the archive does not contain enough information to answer fully, say so plainly and direct the citizen to file a formal follow-up RTI application with the Information Officer specifying the missing detail.

MINISTRY ARCHIVE — VECTOR MATCHES:
%s

MINISTRY ARCHIVE — GRAPH-LINKED PRECEDENT:
%s

RULES:
- Every factual claim must come from the archive shown above. Do not invent figures, names, dates, or document references.
- Cite the relevant prior RTI id (e.g. RTI-2024-0001) or FAQ id when you draw on it.
- Address the citizen directly. Be concise: 4-8 sentences, plain prose, no markdown headings.

OUTPUT FORMAT:
Your reply will be shown to the citizen verbatim. Do NOT include any preamble or signoff like "Response to Citizen:", "Dear Citizen,", or "Best regards". Output ONLY the body of the response, in plain prose, starting with the substantive answer.`

// AnswerRequest drafts a citizen-facing response to a new RTI request.
//
// It assembles the ministry-archive context block from RAG + graph hits and
// asks Claude for a 4-8 sentence answer. If the AI key isn't configured, it
// returns a clearly-labelled stub so the rest of the flow still works in
// local development.
func (c *Client) AnswerRequest(ctx context.Context, subject, description string, ragIdx *rag.Index, graphState *graph.State) (string, error) {
	if !c.Configured() {
		log.Println("[ai] ANTHROPIC_API_KEY not set; returning stub response")
		return "[AI service not configured: set ANTHROPIC_API_KEY] Your request to the Ministry of Climate Change, Environment and Energy has been received and is pending review.", nil
	}

	query := strings.TrimSpace(subject + "\n" + description)

	const ragK = 4
	const graphK = 3

	var vectorHits []rag.Hit
	if ragIdx != nil {
		vectorHits = ragIdx.Retrieve(query, ragK)
	}

	var graphHits []rag.Hit
	if graphState != nil {
		graphHits = graphState.Retrieve(query, graphK+ragK)
	}
	// Dedupe graph hits against vector hits to keep the prompt compact.
	seen := map[string]struct{}{}
	for _, h := range vectorHits {
		if id, ok := h.Payload["id"].(string); ok {
			seen[id] = struct{}{}
		}
	}
	deduped := make([]rag.Hit, 0, len(graphHits))
	for _, h := range graphHits {
		id, _ := h.Payload["id"].(string)
		if _, dup := seen[id]; dup {
			continue
		}
		deduped = append(deduped, h)
		if len(deduped) >= graphK {
			break
		}
	}

	system := fmt.Sprintf(answerSystemTemplate,
		rag.FormatForPrompt(vectorHits),
		graph.FormatForPrompt(deduped),
	)

	userPrompt := fmt.Sprintf("Subject: %s\nDescription: %s\n\nGround your draft in the ministry archive shown in your instructions and write the response to the citizen now.", subject, description)

	answer, err := c.callMessages(ctx, apiMessageRequest{
		Model:     c.Model,
		MaxTokens: 1024,
		System:    system,
		Messages:  []apiMessage{{Role: "user", Content: userPrompt}},
	})
	if err != nil {
		return "", err
	}
	if strings.TrimSpace(answer) == "" {
		return "", fmt.Errorf("anthropic returned empty answer")
	}
	return answer, nil
}

// ----------------------------------------------------------------------------
// Structuring a fresh request for officer review
// ----------------------------------------------------------------------------

// ProcessRequestStructure asks Claude to produce a JSON analysis of a freshly
// filed request: classification, key questions, completeness score, etc.
// On any failure, a heuristic fallback structure is returned so the rest of
// the request-creation flow can continue.
func (c *Client) ProcessRequestStructure(ctx context.Context, subject, description, departmentID string) map[string]any {
	if !c.Configured() {
		log.Println("[ai] no API key — using fallback request structure")
		return fallbackStructure(subject, description)
	}

	prompt := fmt.Sprintf(`You are an AI assistant helping process Right to Information (RTI) requests in the Maldives.

Analyze the following RTI request and provide a structured analysis in JSON format.

CITIZEN REQUEST:
Subject: %s
Description: %s
Department: %s

Provide your analysis as a JSON object with these fields:

1. "request_type": Classify as "Data Request", "Policy Clarification", "Document Access", "Budget Information", "Procedure Inquiry", or "Other"
2. "key_questions": List 2-4 main questions the citizen is asking
3. "information_sought": List specific data, documents, or information items requested
4. "time_period": Extract any time period mentioned or null
5. "geographic_scope": Extract any geographic scope or null
6. "urgency_indicators": List any time-sensitive aspects (empty array if none)
7. "completeness_score": Rate 0.0 to 1.0 how complete and clear the request is
8. "missing_information": List what additional information would make this request clearer (empty array if complete)
9. "related_policies": List relevant Maldivian laws, policies, or RTI Act provisions (empty array if none obvious)
10. "estimated_complexity": "Simple", "Moderate", or "Complex"
11. "suggested_response_approach": Brief 2-3 sentence suggestion on how the officer should approach responding
12. "relevant_precedents": List any similar types of requests that might have been processed before (empty array if none)

Respond ONLY with the JSON object, no other text.`, subject, description, departmentID)

	// Use sonnet for the structuring task (matches the Python implementation).
	model := os.Getenv("ANTHROPIC_STRUCTURE_MODEL")
	if model == "" {
		model = "claude-3-5-sonnet-20241022"
	}
	raw, err := c.callMessages(ctx, apiMessageRequest{
		Model:     model,
		MaxTokens: 2000,
		Messages:  []apiMessage{{Role: "user", Content: prompt}},
	})
	if err != nil {
		log.Printf("[ai] structuring call failed: %v", err)
		return fallbackStructure(subject, description)
	}

	// Strip optional ```json fences before parsing.
	cleaned := strings.TrimSpace(raw)
	if strings.Contains(cleaned, "```json") {
		cleaned = strings.SplitN(cleaned, "```json", 2)[1]
		cleaned = strings.SplitN(cleaned, "```", 2)[0]
	} else if strings.HasPrefix(cleaned, "```") {
		cleaned = strings.TrimPrefix(cleaned, "```")
		if i := strings.Index(cleaned, "```"); i != -1 {
			cleaned = cleaned[:i]
		}
	}
	cleaned = strings.TrimSpace(cleaned)

	var parsed map[string]any
	if err := json.Unmarshal([]byte(cleaned), &parsed); err != nil {
		log.Printf("[ai] could not parse structured response: %v", err)
		return fallbackStructure(subject, description)
	}
	return parsed
}

func fallbackStructure(subject, description string) map[string]any {
	info := description
	if len(info) > 200 {
		info = info[:200] + "..."
	}
	return map[string]any{
		"request_type":                "Data Request",
		"key_questions":               []any{subject},
		"information_sought":          []any{info},
		"time_period":                 nil,
		"geographic_scope":            nil,
		"urgency_indicators":          []any{},
		"completeness_score":          0.6,
		"missing_information":         []any{"Request requires officer review to determine needed clarifications"},
		"related_policies":            []any{"Right to Information Act (Act No. 1/2014)"},
		"estimated_complexity":        "Moderate",
		"suggested_response_approach": "Review the request details and determine if all necessary information has been provided. Contact the citizen if clarification is needed.",
		"relevant_precedents":         []any{},
	}
}
